File: packages/sdk/src/clients/flags.ts
import { SettlerClient } from "../client";

export interface EvaluationContext {
  userId?: string;
  email?: string;
  ip?: string;
  country?: string;
  [key: string]: unknown;
}

export interface FlagEvaluation {
  flagKey: string;
  value: boolean | string | number | Record<string, unknown>;
  variant?: string;
  reason?: string;
}

export class FlagsClient {
  constructor(private client: SettlerClient) {}

  async evaluate(
    flagKey: string,
    context: EvaluationContext = {},
    defaultValue?: boolean | string | number | Record<string, unknown>
  ): Promise<FlagEvaluation> {
    try {
      return await this.client.request<FlagEvaluation>("POST", "/v1/feature-flags/evaluate", {
        body: { flagKey, context, defaultValue },
      });
    } catch (e) {
      if (defaultValue !== undefined) {
        return {
          flagKey,
          value: defaultValue,
          reason: "error_fallback",
        };
      }
      throw e;
    }
  }
}

File: packages/sdk-go/settler/flags.go
package settler

import "context"

// EvaluationContext represents the context used for flag evaluation.
type EvaluationContext map[string]interface{}

// FlagEvaluation represents the result of evaluating a feature flag.
type FlagEvaluation struct {
	FlagKey string      `json:"flagKey"`
	Value   interface{} `json:"value"`
	Variant string      `json:"variant,omitempty"`
	Reason  string      `json:"reason,omitempty"`
}

// FlagsClient handles feature flag operations.
type FlagsClient struct {
	client *Client
}

// Evaluate evaluates a feature flag for the given context.
//
// If defaultValue is provided and evaluation fails, it will return a fallback
// response with reason "error_fallback".
func (c *FlagsClient) Evaluate(ctx context.Context, flagKey string, context EvaluationContext, defaultValue interface{}) (*FlagEvaluation, error) {
	request := map[string]interface{}{
		"flagKey": flagKey,
		"context": context,
	}
	if defaultValue != nil {
		request["defaultValue"] = defaultValue
	}

	resp, err := c.client.requestWithBase(ctx, c.client.rootBaseURL(), "POST", "/v1/feature-flags/evaluate", request, nil)
	if err != nil {
		if defaultValue != nil {
			return &FlagEvaluation{
				FlagKey: flagKey,
				Value:   defaultValue,
				Reason:  "error_fallback",
			}, nil
		}
		return nil, err
	}

	return parseFlagEvaluation(resp, flagKey, defaultValue), nil
}

func parseFlagEvaluation(data map[string]interface{}, flagKey string, defaultValue interface{}) *FlagEvaluation {
	if data == nil {
		return &FlagEvaluation{
			FlagKey: flagKey,
			Value:   defaultValue,
		}
	}
	parsedKey := getString(data, "flagKey")
	if parsedKey == "" {
		parsedKey = flagKey
	}
	return &FlagEvaluation{
		FlagKey: parsedKey,
		Value:   data["value"],
		Variant: getString(data, "variant"),
		Reason:  getString(data, "reason"),
	}
}
